azure-pipelines.yml

```diff
@@ -116,6 +116,87 @@ stages:
+            Python37:
+              minor_version: '7'
+            Python38:
+              minor_version: '8'
+            Python39:
+              minor_version: '9'
+        services:
+          postgres: postgres
+        variables:
+          DOCKER_TAG: none
+        steps:
+          - bash: |
+              # login to docker first so we die fast if this fails
+              docker login -u ${DOCKER_USER} -p ${DOCKER_TOKEN}
+
+              EPOCH=`date +%s`
+              TAG="3.$(minor_version)_develop_${EPOCH}_${RANDOM}"
+              # We repull the develop branch. We could build the local branch that is already present but we should
+              # prune it's depth since we don't need the git history.
+              # CMD="docker buildx build -f docker/Dockerfile.tests --tag greatexpectations/test:${TAG} --target azure --build-arg PYTHON_VERSION=3.$(minor_version) --build-arg SOURCE=github --build-arg BRANCH=develop ."
+              # Temporarily use local copy of repo
+              CMD="docker buildx build -f docker/Dockerfile.tests --tag greatexpectations/test:${TAG} --target azure --build-arg PYTHON_VERSION=3.$(minor_version) ."
+              echo ${CMD}
+              eval ${CMD}
+              PUSH="docker push greatexpectations/test:${TAG}"
+              echo ${PUSH}
+              eval ${PUSH}
+              echo "Docker image can be found at https://hub.docker.com/r/greatexpectations/test/tags"
+              echo "Pull it locally using: docker pull greatexpectations/test:${TAG}"
+
+              # Set this variable so other steps can read it.
+              echo "##vso[task.setvariable variable=DOCKER_TAG]${TAG}"
+
+            displayName: 'Build and push 3.$(minor_version)'
+            env:
+              DOCKER_USER: $(DOCKER_USER)
+              DOCKER_TOKEN: $(DOCKER_TOKEN)
+
+          - bash: |
+              PULL="docker pull greatexpectations/test:$(DOCKER_TAG)"
+              echo ${PULL}
+              eval ${PULL}
+              mkdir coverage
+              PYTEST="docker run --network=host --mount type=bind,source=${PWD}/coverage,target=/coverage "
+              PYTEST+="-e GE_CLOUD_BASE_URL=$(GE_CLOUD_BASE_URL) -e GE_CLOUD_ACCESS_TOKEN=$(GE_CLOUD_ACCESS_TOKEN) "
+              PYTEST+="-e GE_CLOUD_ORGANIZATION_ID=$(GE_CLOUD_ORGANIZATION_ID) greatexpectations/test:$(DOCKER_TAG) "
+              PYTEST+="/bin/bash -c \"pytest --postgresql --spark --cloud --napoleon-docstrings --no-docker-discovery "
+              # We should be able to specify --cov-report=xml:path but this doesn't work so we write to the default
+              # path and then mv it.
+              PYTEST+="--junitxml=/coverage/junit/test-results.xml --cov=. --cov-report=xml --cov-report=html:/coverage/htmlcov "
+              PYTEST+="--ignore=tests/cli --ignore=tests/integration/usage_statistics "
+              PYTEST+="&& mv coverage.xml /coverage/coverage.xml\""
+              echo ${PYTEST}
+              eval ${PYTEST}
```

Comment:
Would these be easier to grok and maintain if there a couple of invoke tasks, or flags on a single invoke tasks?

Even if the answer is yes, it might be easier to save that abstraction for later.